### PR TITLE
intel-tbb: fix patch for old compilers

### LINKED ISSUE #10395: Installation issue: intel-tbb
I was stuck with the installation of intel-tbb when going over [Package Creation Tutorial](https://spack.readthedocs.io/en/latest/tutorial_packaging.html#dependencies). 

```console
root@n7459:~$ spack install intel-tbb
==> pkgconf is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/pkgconf-1.5.4-ry5yoj2yvm3yl3wfuqmnuzslglleu7aw
==> ncurses is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/ncurses-6.1-35xlequxd7hfj36ygbw4umttsdf5pgw4
==> readline is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/readline-7.0-r7vc4akeqmhnqtm3vqu255uzntdwjyxv
==> gdbm is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/gdbm-1.18.1-7a3bryh6jpoku5yoihcktsjfmac6gspg
==> perl is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/perl-5.26.2-p5qgzhl2ucziya2fefq2lvvyowxnzzzf
==> zlib is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/zlib-1.2.11-uk4hz57lzsxrg5h7dr5qse3ng6nsrwcj
==> openssl is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/openssl-1.1.1-w57qpywb6ceu4r234jnj54cywoakxirj
==> cmake is already installed in $SPACK_ROOT/opt/spack/linux-centos6-x86_64/gcc-7.2.0/cmake-3.13.2-3drpny4v5hxqj23ne7goi6waoewdozxs
==> Installing intel-tbb
==> Searching for binary cache of intel-tbb
==> Warning: No Spack mirrors are currently configured
==> No binary for intel-tbb found: installing from source
==> Using cached archive: $SPACK_ROOT/var/spack/cache/intel-tbb/intel-tbb-2019.2.tar.gz
==> Staging archive: $SPACK_ROOT/var/spack/stage/intel-tbb-2019.2-rezvmbclxhnjg4w24opxrmgeii5zndyj/2019_U2.tar.gz
==> Created stage in $SPACK_ROOT/var/spack/stage/intel-tbb-2019.2-rezvmbclxhnjg4w24opxrmgeii5zndyj
1 out of 1 hunk FAILED -- saving rejects to file build/linux.gcc.inc.rej
==> Patch $SPACK_ROOT/var/spack/repos/builtin/packages/intel-tbb/tbb_gcc_rtm_key.patch failed.
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/patch' '-s' '-p' '0' '-i' '$SPACK_ROOT/var/spack/repos/builtin/packages/intel-tbb/tbb_gcc_rtm_key.patch' '-d' '.'

```
Please advice. Thanks.

### PR DESCRIPTION
**Pull request by odoublewen**
This PR appears to fix #10395 -- I ran into the same patch error when building on a centos6 system.  With the changes in this PR, I was able to build the package. 

Between intel-tbb releases `2019` and `2019_U1` (what Spack calls `2019.1`), there was an [upstream change to  `build/linux.gcc.inc`](https://github.com/intel/tbb/commit/4cebdd9a1c9f0c56c4ddcbbb400447ab2a11fad4#diff-76933eb251f8863acd03337b951b04fbL48) that broke our patch file `tbb_gcc_rtm_key.patch`.

However I think this PR still needs some work.  Here are the two most obvious defects I see:

1. According to the comment, the patch should only be applied "on an OS with an elderly assembler".  I don't know what is meant by "elderly" -- was this supposed to mean anything less than gcc 5?  At any rate, the way the `when` spec is written, I think it applies the patch to any gcc **greater than** 4.8.0.  Which will include modern assemblers,  right?!   Seems like the `:` character should be removed so that it reads `when='%gcc@4.8.0'` or even just `when='%gcc@4'`?

2. What is the correct way to specify that the patch file `tbb_gcc_rtm_key.patch` (which is already in develop) should be applied only up to and including `@2019`?  (The patch file does not work starting with `tbb@2019.1`.)  The lack of consistent version tagging system seems to complicate matters.  For example `@2017:2019` (if that is allowed syntax) would seem to also include `2019.1`, etc, right?

@hartzell 

edit: fix markdown hyperlink

### CODE REVIEW

File: var/spack/repos/builtin/packages/intel-tbb/package.py [hunk 1]
```diff
@@ -68,9 +68,12 @@ class IntelTbb(Package):
     depends_on('cmake@3.0.0:', type='build', when='@2017.0:')
 
     # Deactivate use of RTM with GCC when on an OS with an elderly assembler.
-    patch("tbb_gcc_rtm_key.patch", level=0, when='%gcc@4.8.0: os=rhel6')
-    patch("tbb_gcc_rtm_key.patch", level=0, when='%gcc@4.8.0: os=scientific6')
-    patch("tbb_gcc_rtm_key.patch", level=0, when='%gcc@4.8.0: os=centos6')
+    patch("tbb_gcc_rtm_key_2019U1.patch", level=0, when='@2019.1: %gcc@4.8.0: os=rhel6')
```
Comment (odoublewen) [inline var/spack/repos/builtin/packages/intel-tbb/package.py]: Just to be clear, I think `when='%gcc@4.8.0: os=rhel6'` does not equate with the comment "when on an OS with an elderly assembler."

`gcc@4.8.0:` means every gcc from 4.8.0 up to gcc 9 right?

If so, this still needs to be fixed in this PR, I just wasn't sure of the right way to do it.  (Mainly, what is meant by "elderly")

**Comment by odoublewen** [comment]
FYI @chissg

**Comment by odoublewen** [comment]
According to my reading of ["Version ranges" in the Spack docs](https://spack.readthedocs.io/en/latest/packaging_guide.html#version-ranges),
`@:2019` should apply to up to `2019` but not `2019.1` (and beyond).  I wasn't aware of this when I first opened this PR.  Thanks @hartzell for the pointer to the docs.

And I think `when='%gcc@4.8.0:4.99999'` is what was intended by 'elderly compiler' (as opposed to `when='%gcc@4.8.0:'`)

I'm currently testing the new changes (on centos6 with spack build gcc 4.9.2 and 8.2.0) and will update when complete.

**Comment by odoublewen** [comment]
Having trouble with this.  Seems like the Spack docs have a blind spot for perverse versioning schemes like the one being used for this package.  

2019.4, 2019.3, 2019.2, 2019.1, **2019**, 2018.6, etc

There seems to be no way to write `patch(when='')` such that a patch is applied to `2019.1` but not `2019`.  I say this because  `when='@2019.1:'` results in the patch getting applied to `2019`  -- this seems like a spack bug, no?

Note that I **can** handle in inverse case of applying a patch to `2019` but not `2019.1` simply by specifying 
`when='@:2019.0'`

**Comment by hartzell** [comment]
It looks like this is an instance of #1924 (based on discussion in #1920), and there may not be an answer.  

FYI: @davydden @citibeth @adamjstewart @alalazo @tgamblin -- Any suggestion on how Owen can patch *just* things up-to-and-including `2019`?  In this case it seems like *we're* the ones Roy Batty's talking about, since we've assigned the version numbers....

I poked around a bit (although I'm only 98.2% sure that I have the arg's in the right order to match how the `when=` comparison is done) and couldn't find anything that works.

g.


```python
from spack.version import Version, ver

 def toot(a, b):
     if b.satisfies(a):
         print("{0} is satisfied by {1}".format(a, b))
     else:
         print("{0} can't get no satisfaction from {1}".format(a,b))

 # We want a range which is satisfied by "2019.1" and later but not by
 # "2019".

 toot(ver("2019:"), ver(2018))

 toot(ver("2019:"), ver(2019.1))
 toot(ver("2019:"), ver(2019.0))
 toot(ver("2019:"), ver(2019))

 toot(ver("2019.0:"), ver(2019.1))
 toot(ver("2019.0:"), ver(2019.0))
 toot(ver("2019.0:"), ver(2019))

 toot(ver("2019.0.0:"), ver(2019.1))
 toot(ver("2019.0.0:"), ver(2019.0))
 toot(ver("2019.0.0:"), ver(2019))

 toot(ver("2019"), ver(2019.1))
 toot(ver("2019"), ver(2019.0))
 toot(ver("2019"), ver(2019))
```

```consolesession
[ghartzell@bifx1n03 spack]$ PYTHONPATH=$(pwd)/lib/spack:$(pwd)/lib/spack/external python owen.py
2019: can't get no satisfaction from 2018
2019: is satisfied by 2019.1
2019: is satisfied by 2019.0
2019: is satisfied by 2019
2019.0: is satisfied by 2019.1
2019.0: is satisfied by 2019.0
2019.0: is satisfied by 2019
2019.0.0: is satisfied by 2019.1
2019.0.0: is satisfied by 2019.0
2019.0.0: is satisfied by 2019
2019 is satisfied by 2019.1
2019 is satisfied by 2019.0
2019 is satisfied by 2019
```

**Comment by adamjstewart** [comment]
@hartzell @odoublewen See #8957 (and all of the issues it references). We had a very similar problem with `mpich` in which we wanted to patch `3.2` but not `3.2.1`. If you want to patch up to and including `2019` but not `2019.1`, try `':2019.0'`:
```console
$ spack python
Spack version 0.12.1
Python 3.7.0, Darwin x86_64
>>> from spack.version import *
>>> ver('2018.3').satisfies(ver(':2019.0'))
True
>>> ver('2019').satisfies(ver(':2019.0'))
True
>>> ver('2019.1').satisfies(ver(':2019.0'))
False
```

**Comment by odoublewen** [comment]
@adamjstewart Thank you for the suggestion, but that's not my problem.  As I wrote above, I had already figured out how to patch `2019` but not `2019.1`  --  my problem is that there seems to be no way to do the opposite:  patch `2019.1`  but not `2019`.

This seems like a defect, because  **release** `2019` should not satisfy **requirement** `2019.1` -- because it's a bona fide earlier release.  The argument could be made that we should just be calling it `2019.0` but that wasn't how this package was done.  Would be be easier/acceptable to go back and rename the "non-dotted" releases to "dot-zero" i.e. 2017 -> 2017.0 and 2018 -> 2018.0  and 2019 -> 2019.0 ?

@hartzell thanks for the sandboxing -- last night I was trying to figure out how to do exactly that.

**Comment by adamjstewart** [comment]
@odoublewen I think I've found how to reproduce your bug:
```console
$ spack python
Spack version 0.12.1
Python 3.7.0, Darwin x86_64
>>> from spack.version import *
>>> ver('2019').satisfies(ver('2019.1'))
False
>>> ver('2019').satisfies(ver('2019.1:'))
True
>>> ver('2019').satisfies(ver(':2019.1'))
True
```
So it is possible to only patch 2019.1, but not 2019.1+. I'll add this new bug to #8957.

**Comment by odoublewen** [comment]
The crux of it seems to be illustrated by this example: 

```
$ spack python
Spack version 0.12.1
Python 2.6.6, Linux x86_64
>>> from spack.version import ver
>>> ver(2019).satisfies(ver('2019.1'))
False
>>> ver(2019).satisfies(ver('2019.1:'))
True
```

I wouldn't expect the addition of the colon to make the test turn True.  Is this expected / reasonable behavior?

**Comment by odoublewen** [comment]
You beat me to it!

**Comment by odoublewen** [comment]
OK, as a workaround, I am just specifying each of the `2019.x` versions separately and explicitly.  This can be simplified when a solution to #8957 is worked out.

As a side issue, I noticed an issue with another patch file -- `gcc_generic-pedantic-2019.patch` (added in #10836) which despite its name and  actually doesn't work on 2019 -- it works on 2019.1 and beyond.  Basically it's the same situation as the `tbb_gcc_rtm_key.patch` that got this PR started.  So I have applied a similar solution.

I'm running though the various builds, with different compilers, on my centos6 box, and will update when all builds are done.

**Comment by odoublewen** [comment]
OK, with commit 3f45acd I can now successfully install all of the following versions on centos6 (tested with both gcc 4.8.3 and 8.2.0):  `2018.1  2018.6  2019  2019.1  2019.2  2019.3  2019.4`

`2018` fails with a different cmake error -- probably needs a different patch.  My preference is to consider that under a separate, new issue. But I can keep digging if that is preferable.

**Comment by odoublewen** [comment]
Postmortem note:   In 3f45acd, I had to revert my change to the gcc spec for the patches.  It seems these patches are needed by both new and old gcc compiles, at least on `os=*6` machines.  I may have been misinterpreting the comment "on an OS with an elderly assembler", thinking that if I built a shiny new gcc with spack, I could then use it to build intel-tbb without any patches.  But no, they seem to be needed with any gcc, on centos6.

I believe @chissg wrote the comments about the assembler, which I mention in case they want to chime in with thoughts.

**Pull request merged**